http/util.go

```
@@ -36,29 +37,44 @@ func rateLimitQuotaWrapping(handler http.Handler, core *vault.Core) http.Handler
 			respondError(w, http.StatusInternalServerError, err)
 			return
 		}
-		req := w.(*LogicalResponseWriter).request
+
+		// We don't want to do buildLogicalRequestNoAuth here because, if the
+		// request gets allowed by the quota, the same function will get called
+		// again, which is not desired.
+		path, status, err := buildLogicalPathAndOp(r)
+		if err != nil || status != 0 {
+			respondError(w, status, err)
+			return
+		}
+
 		quotaResp, err := core.ApplyRateLimitQuota(&quotas.Request{
 			Type:          quotas.TypeRateLimit,
-			Path:          req.Path,
-			MountPath:     strings.TrimPrefix(req.MountPoint, ns.Path),
+			Path:          path,
+			MountPath:     strings.TrimPrefix(core.MatchingMount(r.Context(), path), ns.Path),
 			NamespacePath: ns.Path,
 			ClientAddress: parseRemoteIPAddress(r),
 		})
 		if err != nil {
-			core.Logger().Error("failed to apply quota", "path", req.Path, "error", err)
+			core.Logger().Error("failed to apply quota", "path", path, "error", err)
 			respondError(w, http.StatusUnprocessableEntity, err)
 			return
 		}
 
 		if !quotaResp.Allowed {
-			quotaErr := errwrap.Wrapf(fmt.Sprintf("request path %q: {{err}}", req.Path), quotas.ErrRateLimitQuotaExceeded)
+			quotaErr := errwrap.Wrapf(fmt.Sprintf("request path %q: {{err}}", path), quotas.ErrRateLimitQuotaExceeded)
 			respondError(w, http.StatusTooManyRequests, quotaErr)
 
 			if core.Logger().IsTrace() {
-				core.Logger().Trace("request rejected due to lease count quota violation", "request_path", req.Path)
+				core.Logger().Trace("request rejected due to lease count quota violation", "request_path", path)
 			}
 
 			if core.RateLimitAuditLoggingEnabled() {
+				req, _, status, err := buildLogicalRequestNoAuth(core.PerfStandby(), w, r)
+				if err != nil || status != 0 {
+					respondError(w, status, err)
+					return
+				}
+
 				_ = core.AuditLogger().AuditRequest(r.Context(), &logical.LogInput{
```

I noticed that we ignore error in other place that we call this, but I wonder if we should at least log it on the server side (as a WARN maybe?) that this audit was dropped.

---

Done.